File: configuration/src/main/java/com/synopsys/integration/configuration/config/PropertyConfiguration.java
package com.synopsys.integration.configuration.config;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import java.util.Set;
import java.util.function.Predicate;
import java.util.stream.Collectors;

import org.apache.commons.lang3.StringUtils;
import org.jetbrains.annotations.NotNull;
import org.springframework.util.Assert;

import com.synopsys.integration.configuration.config.resolution.NoPropertyResolution;
import com.synopsys.integration.configuration.config.resolution.PropertyResolution;
import com.synopsys.integration.configuration.config.resolution.PropertyResolutionInfo;
import com.synopsys.integration.configuration.config.resolution.SourcePropertyResolution;
import com.synopsys.integration.configuration.config.value.ExceptionPropertyValue;
import com.synopsys.integration.configuration.config.value.NoValuePropertyValue;
import com.synopsys.integration.configuration.config.value.PropertyValue;
import com.synopsys.integration.configuration.config.value.ValuedPropertyValue;
import com.synopsys.integration.configuration.parse.ValueParseException;
import com.synopsys.integration.configuration.property.Property;
import com.synopsys.integration.configuration.property.base.NullableProperty;
import com.synopsys.integration.configuration.property.base.PassthroughProperty;
import com.synopsys.integration.configuration.property.base.TypedProperty;
import com.synopsys.integration.configuration.property.base.ValuedProperty;
import com.synopsys.integration.configuration.source.PropertySource;

public class PropertyConfiguration {
    private final Map<String, PropertyResolution> resolutionCache = new HashMap<>();
    private final Map<String, PropertyValue<?>> valueCache = new HashMap<>();
    private final List<PropertySource> orderedPropertySources;

    public PropertyConfiguration(@NotNull List<PropertySource> orderedPropertySources) {
        this.orderedPropertySources = orderedPropertySources;
    }

    //region
    @NotNull
    public <V, R> Optional<R> getValueOrEmpty(@NotNull NullableProperty<V, R> property) {
        assertPropertyNotNull(property);
        try {
            return getValue(property);
        } catch (InvalidPropertyException e) {
            return Optional.empty();
        }
    }

    @NotNull
    public <V, R> R getValueOrDefault(@NotNull ValuedProperty<V, R> property) {
        assertPropertyNotNull(property);
        try {
            return getValue(property);
        } catch (InvalidPropertyException e) {
            V defaultValue = property.getDefaultValue();
            return property.convertValue(defaultValue);
        }
    }

    @NotNull
    public <V, R> Optional<R> getValue(@NotNull NullableProperty<V, R> property) throws InvalidPropertyException {
        assertPropertyNotNull(property);

        PropertyValue<V> value = valueFromCache(property);
        Optional<ValueParseException> parseException = value.getException();
        Optional<PropertyResolutionInfo> propertyResolutionInfo = value.getResolutionInfo();

        if (value.getValue().isPresent()) {
            return value.getValue()
                .map(property::convertValue);
        } else if (parseException.isPresent() && propertyResolutionInfo.isPresent()) {
            throw new InvalidPropertyException(property.getKey(), propertyResolutionInfo.get().getSource(), parseException.get());
        } else {
            return Optional.empty();
        }
    }

    @NotNull
    public <V, R> R getValue(@NotNull ValuedProperty<V, R> property) throws InvalidPropertyException {
        assertPropertyNotNull(property);
        PropertyValue<V> propertyValue = valueFromCache(property);

        Optional<V> value = propertyValue.getValue();
        Optional<ValueParseException> parseException = propertyValue.getException();
        Optional<PropertyResolutionInfo> propertyResolutionInfo = propertyValue.getResolutionInfo();

        if (value.isPresent()) {
            return value.map(property::convertValue).get();
        } else if (parseException.isPresent() && propertyResolutionInfo.isPresent()) {
            throw new InvalidPropertyException(property.getKey(), propertyResolutionInfo.get().getSource(), parseException.get());
        } else {
            V defaultValue = property.getDefaultValue();
            return property.convertValue(defaultValue);
        }
    }

    @NotNull
    public <V, R> Optional<V> getProvidedParsedValue(@NotNull TypedProperty<V, R> property) throws InvalidPropertyException {
        assertPropertyNotNull(property);

        PropertyValue<V> value = valueFromCache(property);
        Optional<ValueParseException> parseException = value.getException();
        Optional<PropertyResolutionInfo> propertyResolutionInfo = value.getResolutionInfo();

        if (value.getValue().isPresent()) {
            return value.getValue();
        } else if (parseException.isPresent() && propertyResolutionInfo.isPresent()) {
            throw new InvalidPropertyException(property.getKey(), propertyResolutionInfo.get().getSource(), parseException.get());
        } else {
            return Optional.empty();
        }
    }

    public boolean wasKeyProvided(@NotNull String key) {
        Assert.notNull(key, "Must provide a property.");
        return resolveFromCache(key).getResolutionInfo().isPresent();
    }

    public <V, R> boolean wasPropertyProvided(@NotNull TypedProperty<V, R> property) {
        assertPropertyNotNull(property);
        return wasKeyProvided(property.getKey());
    }

    public Optional<String> getPropertySource(@NotNull Property property) {
        assertPropertyNotNull(property);
        return resolveFromCache(property.getKey()).getResolutionInfo().map(PropertyResolutionInfo::getSource);
    }

    public Optional<String> getPropertySource(@NotNull String key) {
        Assert.notNull(key, "You must provide a key");
        return resolveFromCache(key).getResolutionInfo().map(PropertyResolutionInfo::getSource);
    }

    public Optional<String> getPropertyOrigin(@NotNull Property property) {
        assertPropertyNotNull(property);
        return resolveFromCache(property.getKey()).getResolutionInfo().map(PropertyResolutionInfo::getOrigin);
    }

    @NotNull
    public Set<String> getKeys() {
        return orderedPropertySources.stream()
            .map(PropertySource::getKeys)
            .flatMap(Set::stream)
            .collect(Collectors.toSet());
    }

    public <V, R> Optional<ValueParseException> getPropertyException(@NotNull TypedProperty<V, R> property) {
        assertPropertyNotNull(property);
        return valueFromCache(property).getException();
    }

    //endregion

    //region Advanced Usage
    public Optional<String> getRaw(@NotNull Property property) {
        assertPropertyNotNull(property, "Must supply a property get raw values.");
        return resolveKey(property.getKey());
    }

    private Optional<String> resolveKey(String key) {
        PropertyResolution propertyResolution = resolveFromCache(key);
        return propertyResolution.getResolutionInfo().map(PropertyResolutionInfo::getRaw);
    }

    @NotNull
    public Map<String, String> getRaw() {
        return getRaw((String key) -> true);
    }

    @NotNull
    public Optional<String> getRaw(@NotNull String key) {
        Assert.notNull(key, "Must supply a key to get a raw value");
        return resolveFromCache(key).getResolutionInfo().map(PropertyResolutionInfo::getRaw);
    }

    @NotNull
    public Map<String, String> getRaw(@NotNull Set<String> keys) {
        Assert.notNull(keys, "Must supply a set of keys to get raw values");
        return getRaw(keys::contains);
    }

    @NotNull
    public Map<String, String> getRawValueMap(@NotNull Set<Property> properties) {
        return getMaskedRawValueMap(properties, key -> false);
    }

    @NotNull
    public Map<String, String> getMaskedRawValueMap(@NotNull Set<Property> properties, Predicate<String> shouldMask) {
        Map<String, String> rawMap = new HashMap<>();
        for (Property property : properties) {
            String rawKey = property.getKey();
            if (property instanceof PassthroughProperty) {
                getRaw((PassthroughProperty) property).entrySet()
                    .forEach(passThrough -> {
                            String fullPassThroughKey = String.format("%s.%s", rawKey, passThrough.getKey());
                            rawMap.put(fullPassThroughKey, maskValue(fullPassThroughKey, passThrough.getValue(), shouldMask));
                        }
                    );
            } else {
                getRaw(property).ifPresent(rawValue -> rawMap.put(rawKey, maskValue(rawKey, rawValue, shouldMask)));
            }
        }
        return rawMap;
    }

    public String maskValue(String rawKey, String rawValue, Predicate<String> shouldMask) {
        String maskedValue = rawValue;
        if (shouldMask.test(rawKey)) {
            maskedValue = StringUtils.repeat('*', maskedValue.length());
        }
        return maskedValue;
    }

    @NotNull
    public Map<String, String> getRaw(@NotNull Predicate<String> predicate) {
        Assert.notNull(predicate, "Must supply a predicate to get raw keys");

        Set<String> keys = getKeys().stream()
            .filter(predicate)
            .filter(Objects::nonNull)
            .collect(Collectors.toSet());

        Map<String, String> keyMap = new HashMap<>();
        keys.forEach(key -> {
            resolveKey(key).ifPresent(rawValue -> keyMap.put(key, rawValue));
        });
        return keyMap;
    }

    // Takes in a 'passthrough.key' and returns key map (whose keys have that value removed)
    // So value 'passthrough.key.example' is returned as 'example'
    @NotNull
    public Map<String, String> getRaw(@NotNull PassthroughProperty property) {
        assertPropertyNotNull(property, "Must supply a passthrough to get raw keys");

        Map<String, String> rawValues = getRaw((String key) -> key.startsWith(property.getKey()));
        Map<String, String> trimmedKeys = new HashMap<>();
        for (Map.Entry<String, String> entry : rawValues.entrySet()) {
            if (entry.getValue() != null) {
                trimmedKeys.put(property.trimKey(entry.getKey()), entry.getValue());
            }
        }
        return trimmedKeys;
    }
    //endregion Advanced Usage

    //region Implementation Details

    private void assertPropertyNotNull(Property property) {
        assertPropertyNotNull(property, "Must provide a property.");
    }

    private void assertPropertyNotNull(Property property, @NotNull String message) {
        Assert.notNull(property, message);
    }

    private PropertyResolution resolveFromCache(@NotNull String key) {
        Assert.notNull(key, "Cannot resolve a null key.");
        if (!resolutionCache.containsKey(key)) {
            resolutionCache.put(key, resolveFromPropertySources(key));
        }

        PropertyResolution value = resolutionCache.get(key);
        Assert.notNull(value, "Could not resolve a value, something has gone wrong with properties!");
        return value;
    }

    private PropertyResolution resolveFromPropertySources(@NotNull String key) {
        Assert.notNull(key, "Cannot resolve a null key.");
        for (PropertySource propertySource : orderedPropertySources) {
            if (propertySource.hasKey(key)) {
                String rawValue = propertySource.getValue(key);
                if (rawValue != null) {
                    String name = propertySource.getName();
                    String origin = propertySource.getOrigin(key);
                    PropertyResolutionInfo propertyResolutionInfo = new PropertyResolutionInfo(name, origin, rawValue);
                    return new SourcePropertyResolution(propertyResolutionInfo);
                }
            }
        }
        return new NoPropertyResolution();
    }

    @NotNull
    private <V, R> PropertyValue<V> valueFromCache(@NotNull TypedProperty<V, R> property) {
        if (!valueCache.containsKey(property.getKey())) {
            valueCache.put(property.getKey(), valueFromResolution(property));
        }

        PropertyValue<V> value = (PropertyValue<V>) valueCache.get(property.getKey());
        Assert.notNull(value, "Could not source a value, something has gone wrong with properties!");
        return value;
    }

    @NotNull
    private <V, R> PropertyValue<V> valueFromResolution(@NotNull TypedProperty<V, R> property) {
        Assert.notNull(property, "Cannot resolve a null property.");
        Optional<PropertyResolutionInfo> propertyResolution = resolveFromCache(property.getKey()).getResolutionInfo();
        return propertyResolution
            .map(propertyResolutionInfo -> coerceValue(property, propertyResolutionInfo))
            .orElseGet(NoValuePropertyValue::new);
    }

    @NotNull
    private <V, R> PropertyValue<V> coerceValue(@NotNull TypedProperty<V, R> property, @NotNull PropertyResolutionInfo propertyResolutionInfo) {
        Assert.notNull(property, "Cannot resolve a null property.");
        Assert.notNull(propertyResolutionInfo, "Cannot coerce a null property resolution.");
        try {
            V value = property.getValueParser().parse(propertyResolutionInfo.getRaw());
            return new ValuedPropertyValue<>(value, propertyResolutionInfo);
        } catch (ValueParseException e) {
            return new ExceptionPropertyValue<>(e, propertyResolutionInfo);
        }
    }
    //endregion
}

File: configuration/src/main/java/com/synopsys/integration/configuration/config/PropertyConfiguration.kt
/**
 * configuration
 *
 * Copyright (c) 2020 Synopsys, Inc.
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package com.synopsys.integration.configuration.config

import com.synopsys.integration.configuration.parse.ValueParseException
import com.synopsys.integration.configuration.property.Property
import com.synopsys.integration.configuration.property.base.NullableProperty
import com.synopsys.integration.configuration.property.base.PassthroughProperty
import com.synopsys.integration.configuration.property.base.TypedProperty
import com.synopsys.integration.configuration.property.base.ValuedProperty

class PropertyConfiguration(private val orderedPropertySources: List<PropertySource>) {
    private val resolutionCache: MutableMap<String, PropertyResolution> = mutableMapOf()
    private val valueCache: MutableMap<String, PropertyValue> = mutableMapOf()

    //#region Recommended Usage
    fun <T> getValueOrNull(property: NullableProperty<T>): T? {
        return try {
            getValue(property)
        } catch (e: InvalidPropertyException) {
            null
        }
    }

    fun <T> getValueOrDefault(property: ValuedProperty<T>): T {
        return try {
            getValue(property)
        } catch (e: InvalidPropertyException) {
            property.default
        }
    }

    @Throws(InvalidPropertyException::class)
    @Suppress("UNCHECKED_CAST")
    fun <T> getValue(property: NullableProperty<T>): T? {
        return when (val value = valueFromCache(property)) {
            is TypedValue<*> -> value.value as T
            is ExceptionValue -> throw InvalidPropertyException(property.key, value.resolution.source, value.exception)
            is NoValue -> null
        }
    }

    @Throws(InvalidPropertyException::class)
    @Suppress("UNCHECKED_CAST")
    fun <T> getValue(property: ValuedProperty<T>): T {
        return when (val value = valueFromCache(property)) {
            is TypedValue<*> -> value.value as T
            is ExceptionValue -> throw InvalidPropertyException(property.key, value.resolution.source, value.exception)
            is NoValue -> property.default
        }
    }

    fun wasKeyProvided(key: String): Boolean {
        return when (resolveFromCache(key)) {
            is SourceResolution -> true
            is NoResolution -> false
        }
    }

    fun <T> wasPropertyProvided(property: TypedProperty<T>): Boolean {
        return wasKeyProvided(property.key)
    }

    fun getPropertySource(property: Property): String? {
        return when (val value = resolveFromCache(property.key)) {
            is SourceResolution -> value.source
            is NoResolution -> null
        }
    }

    fun getPropertyOrigin(property: Property): String? {
        return when (val value = resolveFromCache(property.key)) {
            is SourceResolution -> value.origin
            is NoResolution -> null
        }
    }

    fun getKeys(): Set<String> {
        return orderedPropertySources.flatMap { it.getKeys() }.toSet()
    }

    fun <T> getPropertyException(property: TypedProperty<T>): ExceptionValue? {
        return when (val value = valueFromCache(property)) {
            is ExceptionValue -> value
            else -> null
        }
    }

    //#endregion Recommended Usage

    //region Advanced Usage
    fun getRaw(property: Property): String? {
        return when (val value = resolveFromCache(property.key)) {
            is NoResolution -> null
            is SourceResolution -> value.raw
        }
    }

    fun getRaw(): Map<String, String> {
        return getRaw { true }
    }

    fun getRaw(keys: Set<String>): Map<String, String> {
        return getRaw { key -> keys.contains(key) }
    }

    fun getRaw(matches: (String) -> Boolean): Map<String, String> {
        return getKeys()
                .filter(matches::invoke)
                .mapNotNull { key ->
                    when (val resolution = resolveFromCache(key)) {
                        is SourceResolution -> Pair(key, resolution.raw)
                        is NoResolution -> null
                    }
                }.toMap()
    }

    // Takes in a 'passthrough.key' and returns key map (whose keys have that value removed)
    // So value 'passthrough.key.example' is returned as 'example'
    fun getRaw(property: PassthroughProperty): Map<String, String> {
        return getRaw { key -> key.startsWith(property.key) }
                .map { pair -> pair.key.substring(property.key.length + 1) to pair.value }
                .toMap()
    }
    //endregion Advanced Usage

    //region Implementation Details
    private fun resolveFromCache(key: String): PropertyResolution {
        if (!resolutionCache.containsKey(key)) {
            resolutionCache[key] = resolveFromPropertySources(key)
        }

        return resolutionCache[key] ?: throw RuntimeException("Could not resolve a value, something has gone wrong with properties!")
    }

    private fun resolveFromPropertySources(key: String): PropertyResolution {
        for (source in orderedPropertySources) {
            if (source.hasKey(key)) {
                val rawValue = source.getValue(key)
                if (rawValue != null) { // If this property source is the first with a value, it is the canonical source of this property key.
                    val propertySourceName = source.getName()
                    val propertyOrigin = source.getOrigin(key) ?: "unknown"
                    return SourceResolution(propertySourceName, propertyOrigin, rawValue)
                }
            }
        }
        return NoResolution
    }

    private fun <T> valueFromCache(property: TypedProperty<T>): PropertyValue {
        if (!valueCache.containsKey(property.key)) {
            valueCache[property.key] = valueFromResolution(property)
        }

        return valueCache[property.key] ?: throw RuntimeException("Could not resolve a value, something has gone wrong with properties!")
    }

    private fun <T> valueFromResolution(property: TypedProperty<T>): PropertyValue {
        return when (val resolution = resolveFromCache(property.key)) {
            is NoResolution -> NoValue
            is SourceResolution -> coerceValue(property, resolution)
        }
    }

    private fun <T> coerceValue(property: TypedProperty<T>, resolution: SourceResolution): PropertyValue {
        return try {
            val value = property.parser.parse(resolution.raw)
            TypedValue(value, resolution)
        } catch (e: ValueParseException) {
            ExceptionValue(e, resolution)
        }
    }
    //endregion Implementation Details
}

class InvalidPropertyException(propertyKey: String, propertySourceName: String, innerException: ValueParseException) : Exception(
        "The key '${propertyKey}' in property source '${propertySourceName}' contained a value that could not be reasonably converted to the properties type. The exception was: ${innerException.localizedMessage ?: "Unknown"}",
        innerException
) {}

sealed class PropertyValue
data class TypedValue<T>(val value: T, val resolution: SourceResolution) : PropertyValue() // A property source contained a value and the value could be parsed to the proper type.
data class ExceptionValue(val exception: ValueParseException, val resolution: SourceResolution) : PropertyValue() // A property source contained a value but the value could NOT be parsed to the proper type.
object NoValue : PropertyValue()

sealed class PropertyResolution
data class SourceResolution(val source: String, val origin: String, val raw: String) : PropertyResolution()
object NoResolution : PropertyResolution() // No property source contained a value.